math/src/elliptic_curve/short_weierstrass/curves/bls12_381/pairing.rs

```diff
@@ -42,10 +109,53 @@ impl IsPairing for BLS12381AtePairing {
                 result *= miller(&q, &p);
             }
         }
-        Ok(final_exponentiation(&result))
+        Ok(final_exponentiation_optimized(&result))
     }
 }
 
+/// Implements the miller loop for the ate pairing of the BLS12 381 curve.
+/// Based on algorithm 9.2, page 212 of the book
+/// "Topics in computational number theory" by W. Bons and K. Lenstra
+#[allow(unused)]
+pub fn miller(
+    q: &ShortWeierstrassProjectivePoint<BLS12381TwistCurve>,
+    p: &ShortWeierstrassProjectivePoint<BLS12381Curve>,
+) -> FieldElement<Degree12ExtensionField> {
+    let mut r = q.clone();
+    let mut f = FieldElement::<Degree12ExtensionField>::one();
+    let mut miller_loop_constant = MILLER_LOOP_CONSTANT;
+    let mut miller_loop_constant_bits: alloc::vec::Vec<bool> = alloc::vec![];
+
+    while miller_loop_constant > 0 {
+        miller_loop_constant_bits.insert(0, (miller_loop_constant & 1) == 1);
+        miller_loop_constant >>= 1;
+    }
+
+    for bit in miller_loop_constant_bits[1..].iter() {
+        double_accumulate_line(&mut r, p, &mut f);
+        if *bit {
+            add_accumulate_line(&mut r, q, p, &mut f);
+        }
+    }
+    f.conjugate()
+}
+#[allow(unused)]
+pub fn miller_optimized(
+    q: &ShortWeierstrassProjectivePoint<BLS12381TwistCurve>,
+    p: &ShortWeierstrassProjectivePoint<BLS12381Curve>,
+) -> FieldElement<Degree12ExtensionField> {
+    let mut r = q.clone();
+    let mut f = FieldElement::<Degree12ExtensionField>::one();
+    X_BINARY.iter().skip(1).for_each(|bit| {
+        double_accumulate_line(&mut r, p, &mut f);
+        if *bit {
+            add_accumulate_line(&mut r, q, p, &mut f);
+        }
+    });
+
+    f.conjugate()
+}
```

Comment:
Maybe we can just keep `miller_optimized` as `miller` and, if we need the old one for validation, call it `miller_slow`.